### PR TITLE
[SPARK-43926][CONNECT][PYTHON] Add array_agg, array_size, cardinality, count_min_sketch,mask,named_struct,json_* to Scala and Python

### PR DESCRIPTION
**Pull request by ivoson**
### What changes were proposed in this pull request?

Add following functions:
- array_agg
- array_size
- cardinality
- count_min_sketch
- named_struct
- json_array_length
- json_object_keys
- mask

To:

- Scala API
- Python API
- Spark Connect Scala Client
- Spark Connect Python Client

### Why are the changes needed?
Add Scala, Python and Connect API for these sql functions: array_agg, array_size, cardinality, count_min_sketch, named_struct, json_array_length, json_object_keys, mask

### Does this PR introduce _any_ user-facing change?
Yes, added new functions.

### How was this patch tested?
New UT added.

### CODE REVIEW

**Comment by ivoson** [comment]
cc @zhengruifeng

**Comment by LuciferYang** [comment]
also cc @HyukjinKwon @panbingkun @beliefer FYI

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -6379,6 +6428,32 @@ object functions {
...
+   * @group string_funcs
+   * @since 3.5.0
+   */
+  // scalastyle:on line.size.limit
+  def mask(
+    input: Column,
+    upperChar: Column,
+    lowerChar: Column,
+    digitChar: Column,
+    otherChar: Column): Column = {
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:6451]: Please supplement the API with other constructor of `Mask`.

Comment (ivoson) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:6451]: Sure, will add the other constuctors. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -368,6 +368,24 @@ object functions {
    */
   def collect_set(columnName: String): Column = collect_set(Column(columnName))
 
+  /**
```
Comment (panbingkun) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:371]: Returns a count-min sketch of a column with the given esp,
      confidence and seed. The result is an array of bytes, which can be deserialized to a
      `CountMinSketch` before usage. Count-min sketch is a probabilistic data structure used for
      cardinality estimation using sub-linear space.

Is the comment above more aligned?

https://github.com/apache/spark/blob/00e7c08606d0b6de22604d2a7350ea0711355300/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CountMinSketchAgg.scala#L45-L48

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -14394,6 +14394,260 @@ def nvl2(col1: "ColumnOrName", col2: "ColumnOrName", col3: "ColumnOrName") -> Co
     return _invoke_function_over_columns("nvl2", col1, col2, col3)
 
 
+@try_remote_functions
```
Comment (panbingkun) [inline python/pyspark/sql/functions.py:14397]: Suggest placing the function in the appropriate location of the corresponding file

File: python/pyspark/sql/connect/functions.py [hunk 4]
```diff
@@ -3603,6 +3603,80 @@ def nvl2(col1: "ColumnOrName", col2: "ColumnOrName", col3: "ColumnOrName") -> Co
 nvl2.__doc__ = pysparkfuncs.nvl2.__doc__
```
Comment (panbingkun) [inline python/pyspark/sql/connect/functions.py:3605]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 5]
```diff
@@ -5537,6 +5568,61 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
     )
   }
 
+  test("json_array_length function") {
```
Comment (panbingkun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:5571]: Can these tests be placed in `JsonFunctionsSuite`?

File: python/pyspark/sql/functions.py [hunk 6]
```diff
@@ -14394,6 +14394,260 @@ def nvl2(col1: "ColumnOrName", col2: "ColumnOrName", col3: "ColumnOrName") -> Co
...
+    [Row(r='YyYYddd-@$#'), Row(r='yyyy-YYYY-dddd-dddd')]
+    >>> df.select(mask(df.data, lit('Y'), lit('y'), lit('d'), lit('*')).alias('r')).collect()
+    [Row(r='YyYYddd****'), Row(r='yyyy*YYYY*dddd*dddd')]
+    """
+
+    _upperChar = lit("X") if upperChar is None else upperChar
+    _lowerChar = lit("x") if lowerChar is None else lowerChar
+    _digitChar = lit("n") if digitChar is None else digitChar
+    _otherChar = lit(None) if otherChar is None else otherChar
+    return _invoke_function_over_columns(
```
Comment (beliefer) [inline python/pyspark/sql/functions.py:14646]: If user passed lowerChar without upperChar, what will happens?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 2 above)
Comment (ivoson) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:371]: Thanks, done.

File: python/pyspark/sql/functions.py (hunk 3 above)
Comment (ivoson) [reply python/pyspark/sql/functions.py]: relocate the functions by group, thanks.

File: python/pyspark/sql/connect/functions.py (hunk 4 above)
Comment (ivoson) [reply python/pyspark/sql/connect/functions.py]: thanks, done

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 5 above)
Comment (ivoson) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: thanks, moved to `JsonFunctionsSuite`

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 7]
```diff
@@ -6379,6 +6428,100 @@ object functions {
...
+   */
+  // scalastyle:on line.size.limit
+  def mask(input: Column, upperChar: Column): Column = withExpr {
+    new Mask(input.expr, upperChar.expr)
+  }
+
+  // scalastyle:off line.size.limit
+  /**
+   * Masks the given string value. The function replaces upper-case, lower-case characters
+   * with specific characters, and numbers with 'n'.
```
[the remark this answers is not in the document]
Comment (ivoson) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 8]
```diff
@@ -6379,6 +6428,100 @@ object functions {
...
+   */
+  // scalastyle:on line.size.limit
+  def mask(input: Column, upperChar: Column, lowerChar: Column): Column = withExpr {
+    new Mask(input.expr, upperChar.expr, lowerChar.expr)
+  }
+
+  // scalastyle:off line.size.limit
+  /**
+   * Masks the given string value. The function replaces upper-case, lower-case characters and
+   * numbers with specific characters.
```
[the remark this answers is not in the document]
Comment (ivoson) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Thanks, done.

File: python/pyspark/sql/functions.py (hunk 6 above)
Comment (ivoson) [reply python/pyspark/sql/functions.py]: In this case, will repace lowerChar with specified character, and other cases with default value.

For example,
```
mask("col", lowerChar = 'y')
```
will replace:
lowerChar -> 'y'
upperChar -> 'X'
digitChar -> 'n'
otherChar -> retain original

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 9]
```diff
@@ -4454,6 +4486,106 @@ object functions {
...
+   * removed.
+   *
+   * @param input
+   *   string value to mask. Supported types: STRING, VARCHAR, CHAR
+   * @group string_funcs
+   * @since 3.5.0
+   */
+  def mask(input: Column): Column = Column.fn("mask", input)
+
+  // scalastyle:off line.size.limit
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:4501]: I think we should fix the code style and remove these `scalastyle:off line.size.limit`

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 10]
```diff
@@ -368,6 +368,23 @@ object functions {
...
+  /**
+   * Returns a count-min sketch of a column with the given esp, confidence and seed. The result
+   * is an array of bytes, which can be deserialized to a `CountMinSketch` before usage.
+   * Count-min sketch is a probabilistic data structure used for cardinality estimation using
+   * sub-linear space.
+   *
+   * @group agg_funcs
+   * @since 3.5.0
+   */
+  def count_min_sketch(
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:380]: What's the difference between this one and `DataFrameStatFunctions#countMinSketch`?

File: python/pyspark/sql/functions.py (hunk 6 above)
Comment (beliefer) [reply python/pyspark/sql/functions.py]: I mean is `mask("col", 'y')`, then user will be confused.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 11]
```diff
@@ -6671,6 +6719,100 @@ object functions {
...
+   *
+   * @param input string value to mask. Supported types: STRING, VARCHAR, CHAR
+   * @group string_funcs
+   * @since 3.5.0
+   */
+  def mask(input: Column): Column = withExpr {
+    new Mask(input.expr)
+  }
+
+  // scalastyle:off line.size.limit
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:6735]: ditto, we should avoid to use `scalastyle:off` for the new code

File: python/pyspark/sql/functions.py (hunk 6 above)
Comment (beliefer) [reply python/pyspark/sql/functions.py]: According to the discussion between @zhengruifeng and me, the analyzer will check it too. You can temporarily ignore the above comment now.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 9 above)
Comment (zhengruifeng) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:4501]: +1, please remove them,

I know there are some usage of `scalastyle:off line.size.limit` in this file, but they are due to long hyperlink string which can not be broken into multi line.

In this PR, i think it is not needed

Comment (ivoson) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala]: Done. thanks @LuciferYang @zhengruifeng

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 11 above)
Comment (ivoson) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: got it. thanks.

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
thanks, merged to master

**Comment by ivoson** [comment]
thanks a lot for your review. @zhengruifeng @LuciferYang @beliefer @panbingkun

**Pull request closed without merging**